File: built-in plugins dsp/mb-master-k-weight.cpp
/**
 * MB K-Weighted Meter
 * Category : effect
 * Type     : mastering
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : K-system metering for calibrated monitoring and mastering
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MASTER_K_WEIGHT_H
#define MB_MASTER_K_WEIGHT_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMasterKWeight : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-master-k-weight";
    static constexpr const char* PLUGIN_NAME    = "MB K-Weighted Meter";
    static constexpr const char* PLUGIN_TYPE    = "mastering";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float kScale = 14f;  // range [12, 20]
    float holdTime = 2000f;  // range [500, 5000]
    float fallRate = 20f;  // range [5, 50]
    };

    MbMasterKWeight() = default;
    ~MbMasterKWeight() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.kScale = std::clamp(params.kScale, 12f, 20f);
        params.holdTime = std::clamp(params.holdTime, 500f, 5000f);
        params.fallRate = std::clamp(params.fallRate, 5f, 50f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB K-Weighted Meter
        return input;
    }
};

#endif // MB_MASTER_K_WEIGHT_H

File: built-in plugins dsp/mb-master-k-weight.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbMasterKWeightPlugin: PluginDefinition = {
  id: "mb-master-k-weight",
  slug: "mb-master-k-weight",
  name: "MB K-Weighted Meter",
  category: "effect",
  type: "mastering" as any,
  version: "1.0.0",
  description: "K-system metering for calibrated monitoring and mastering",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "kScale",
      name: "K Scale",
      type: "float",
      defaultValue: 14,
      minValue: 12,
      maxValue: 20,
      automatable: false,
    },
    {
      id: "holdTime",
      name: "Hold Time",
      type: "float",
      defaultValue: 2000,
      minValue: 500,
      maxValue: 5000,
      automatable: false,
    },
    {
      id: "fallRate",
      name: "Fall Rate",
      type: "float",
      defaultValue: 20,
      minValue: 5,
      maxValue: 50,
      automatable: false,
    },
  ],
  defaultPreset: { kScale: 14, holdTime: 2000, fallRate: 20 },
};

export default MbMasterKWeightPlugin;
